docs/flip/1009-optimizer-api.md

```diff
@@ -0,0 +1,541 @@
+- Start Date: 2021-02-08
+- FLIP PR: [#1011](https://github.com/google/flax/pull/1011)
+- FLIP Issue: [#1009](https://github.com/google/flax/issues/1009)
+
+Table of contents:
+
+- [Summary]
+- [Motivation]
+- [Functional API]
+  - [Functional Implementation]
+  - [Using Optax]
+  - [Multi Optimizer]
+  - [Functional Usage]
+  - [Linen Helper]
+- [Previous API]
+  - [Previous Implementation]
+  - [Previous Usage]
+- [Update Plan]
+
+
+# Summary
+[Summary]: #summary
+
+Our current API (referred to as [previous API] in this document) uses a pattern
+where an `Optimizer` dataclass is created from a pytree of `target` variables
+and from an `OptimizerDef` that defines how to update optimizer state,
+hyperparameters, and target variables. This pattern is relatively complex for
+implementing a simple optimizer, while being quite verbose in the typical Linen
+train step (especially when using mutable state collections).
+
+This FLIP proposes a new purely [functional API] where optimizers are simply
+defined by an `init()` and an `update()` function. This new pattern allows
+[using Optax] optimizers that follow a very similar functional pattern.
+
+In the end, this document also proposes a new small [Linen helper] that takes
+care of the usual splitting and merging of variables for optimization.
+
+
+# Motivation
+[motivation]: #motivation
+
+Flax contains some optimizers in the `flax.optim` package, but that list is far
+from exhaustive and ideally we would instead use JAX optimizers from a dedicated
+PyPi package.
+
+DeepMind already has a dedicated library — [Optax] — that implements a wide
+range of interesting optimizers and provides a framework to compose new
+optimizers from reusable gradient transformations.
+
+[Optax]: https://github.com/deepmind/optax
+
+
+# Functional API
+[Functional API]: #functional-api
+
+## Functional Implementation
+[Functional Implementation]: #functional-implementation
+
+An optimizer is defined by two pure functions. Since we will be [using Optax]
+we define the `Optimizer` as a named tuple (instead of a
+`flax.struct.dataclass`):
+
+```python
+# flax/opt/base.py
+
+Params = Any
+Gradients = Any
+OptimizerState = NamedTuple
+
+
+class Optimizer(NamedTuple):
+  init: Callable[[Params], OptimizerState]
+  update: Callable[
+      [Gradients, Params, OptimizerState], Tuple[Params, OptimizerState]]
+```
+
+For example, to implement `flax.optim.Momentum` in this new interface:
+
+```python
+# flax/opt/momentum.py
+
+class MomentumState(OptimizerState):
+  momentum: PyTree
+
+
+def momentum(learning_rate: float, beta: float = 0.9) -> Optimizer:
```

Comment:
I think it's good to discuss the pros/cons of this construction pattern a bit more. I think there are some downsides to the higher-order function pattern here:
- hyper params are not inspectable
- requires static_argnums= to pass into Jax functions
- you cannot implement auxiliary functions like `__repr__` or other utility like methods beside init/update

Reply:
Thanks for highlighting these points...

- inspect construction params : added comment to discussion under `Gradient Transformations`
- inspect hyper parameters that are scheduled : see comments about `learning_rate` in discussion `Optax Training Step`
- wrt `static_argnums` : solved with the proposed `Train State` helper
- in principle `OptState` could be extended with utility functions shared between all optimizers, and individual `OptState` can be freely extended ...